File: packages/server/src/lib/cache.ts
import type { ActiveTask, CompletedTaskMeta } from "./types.js";
import type { GatewayWorkerInspect } from "./celery-gateway.js";
import { gatewayGetActiveTasks } from "./celery-gateway.js";
import { getActiveTasks, getHistoricalTasks, getKnownTaskNames, getQueueLengths, getPendingTasks } from "./celery.js";
import { gatewayGetQueues, gatewayGetRegisteredTasks, gatewayInspectWorkers } from "./celery-gateway.js";
import { getRedis } from "./redis.js";

// ---------------------------------------------------------------------------
// Generic cache infrastructure
// ---------------------------------------------------------------------------

interface CacheEntry<T = unknown> {
  data: T;
  updatedAt: number;
  ttlMs: number;
  timer: ReturnType<typeof setInterval> | null;
  inflight: Promise<T> | null;
  refreshFn: () => Promise<T>;
}

class CeleryCache {
  private entries = new Map<string, CacheEntry>();

  constructor() {
    this.register("active-tasks", 2_000, refreshActiveTasks);
    this.register("queue-depths", 30_000, refreshQueueDepths);
    this.register("task-history", 10_000, refreshTaskHistory);
    this.register("worker-inspect", 15_000, refreshWorkerInspect);
    this.register("registered-tasks", 60_000, refreshRegisteredTasks);
    this.register("queue-details", 5_000, refreshQueueDetails);
  }

  private register<T>(
    key: string,
    ttlMs: number,
    refreshFn: () => Promise<T>,
  ): void {
    this.entries.set(key, {
      data: undefined as T,
      updatedAt: 0,
      ttlMs,
      timer: null,
      inflight: null,
      refreshFn,
    } as CacheEntry);
  }

  /**
   * Read cached data. First call awaits the initial fetch and starts the
   * background timer. Subsequent calls return instantly from memory.
   */
  async get<T>(key: string): Promise<T> {
    const entry = this.entries.get(key) as CacheEntry<T> | undefined;
    if (!entry) throw new Error(`Unknown cache key: ${key}`);

    // Cold start – first access
    if (entry.updatedAt === 0) {
      await this.refresh(key);
      this.startTimer(key);
    }

    return entry.data;
  }

  /**
   * Execute the refresh function. Deduplicates concurrent calls and serves
   * stale data on error.
   */
  private async refresh(key: string): Promise<void> {
    const entry = this.entries.get(key)!;

    if (entry.inflight) {
      await entry.inflight.catch(() => {});
      return;
    }

    const promise = entry.refreshFn();
    entry.inflight = promise;

    try {
      const data = await promise;
      entry.data = data;
      entry.updatedAt = Date.now();
    } catch (err) {
      console.warn(`[cache] refresh failed for "${key}":`, err);
    } finally {
      entry.inflight = null;
    }
  }

  private startTimer(key: string): void {
    const entry = this.entries.get(key)!;
    if (entry.timer) return;

    entry.timer = setInterval(() => {
      this.refresh(key).catch(() => {});
    }, entry.ttlMs);

    // Don't keep the Node process alive just for cache timers
    if (entry.timer && typeof entry.timer === "object" && "unref" in entry.timer) {
      entry.timer.unref();
    }
  }
}

// ---------------------------------------------------------------------------
// Singleton (simple module-level variable)
// ---------------------------------------------------------------------------

let cache: CeleryCache | null = null;

function getCache(): CeleryCache {
  if (!cache) cache = new CeleryCache();
  return cache;
}

// ---------------------------------------------------------------------------
// Refresh functions – absorb the gateway-first / Redis-fallback logic
// ---------------------------------------------------------------------------

async function refreshActiveTasks(): Promise<ActiveTask[]> {
  try {
    const { tasks } = await gatewayGetActiveTasks();
    return tasks.map((t) => ({
      taskId: t.id,
      name: t.name,
      worker: t.worker,
      startedAt: t.time_start ?? Date.now() / 1000,
      status: t.acknowledged ? ("started" as const) : ("received" as const),
      args: t.args ?? undefined,
      kwargs: t.kwargs ?? undefined,
    }));
  } catch {
    try {
      return await getActiveTasks();
    } catch {
      return [];
    }
  }
}

async function refreshQueueDepths(): Promise<Record<string, number>> {
  try {
    const { queues } = await gatewayGetQueues();
    const result: Record<string, number> = {};
    for (const q of queues) {
      result[q.name] = q.depth;
    }
    return result;
  } catch {
    try {
      const depths = await getQueueLengths(["celery"]);
      const result: Record<string, number> = {};
      for (const [key, value] of depths) {
        result[key] = value;
      }
      return result;
    } catch {
      return { celery: 0 };
    }
  }
}

async function refreshTaskHistory(): Promise<CompletedTaskMeta[]> {
  try {
    const tasks = await getHistoricalTasks(50);

    // Backfill known task names (side effect preserved from original route)
    const names = tasks
      .map((t) => t.name)
      .filter((n) => n && n !== "unknown");
    if (names.length > 0) {
      getRedis().sadd("celeryhub:known-tasks", ...names).catch(() => {});
    }

    return tasks;
  } catch {
    return [];
  }
}

async function refreshWorkerInspect(): Promise<GatewayWorkerInspect | null> {
  try {
    return await gatewayInspectWorkers();
  } catch {
    return null;
  }
}

export interface RegisteredTasksResult {
  byWorker: Record<string, string[]>;
  tasks: string[];
}

async function refreshRegisteredTasks(): Promise<RegisteredTasksResult> {
  try {
    const knownNames = await getKnownTaskNames();

    let byWorker: Record<string, string[]> = {};
    try {
      const registered = await gatewayGetRegisteredTasks();
      byWorker = registered.by_worker;
    } catch {
      // Gateway unavailable – still have persistent names
    }

    const allTasks = new Set<string>(knownNames);
    for (const tasks of Object.values(byWorker)) {
      for (const t of tasks) allTasks.add(t);
    }

    return { byWorker, tasks: Array.from(allTasks).sort() };
  } catch {
    return { byWorker: {}, tasks: [] };
  }
}

export interface QueueDetailsResult {
  queueNames: string[];
  depths: Record<string, number>;
  pending: Record<
    string,
    { taskId: string; taskName: string; enqueuedAt: string }[]
  >;
}

async function refreshQueueDetails(): Promise<QueueDetailsResult> {
  const c = getCache();

  let depths: Record<string, number>;
  try {
    depths = await c.get<Record<string, number>>("queue-depths");
  } catch {
    depths = { celery: 0 };
  }

  const queueNames = Object.keys(depths);
  const pending: QueueDetailsResult["pending"] = {};

  await Promise.all(
    queueNames.map(async (q) => {
      try {
        pending[q] = await getPendingTasks(q, 20);
      } catch {
        pending[q] = [];
      }
    }),
  );

  return { queueNames, depths, pending };
}

// ---------------------------------------------------------------------------
// Public typed convenience accessors
// ---------------------------------------------------------------------------

export async function getCachedActiveTasks(): Promise<ActiveTask[]> {
  return getCache().get<ActiveTask[]>("active-tasks");
}

export async function getCachedQueueDepths(): Promise<Record<string, number>> {
  return getCache().get<Record<string, number>>("queue-depths");
}

export async function getCachedTaskHistory(): Promise<CompletedTaskMeta[]> {
  return getCache().get<CompletedTaskMeta[]>("task-history");
}

export async function getCachedWorkerInspect(): Promise<GatewayWorkerInspect | null> {
  return getCache().get<GatewayWorkerInspect | null>("worker-inspect");
}

export async function getCachedRegisteredTasks(): Promise<RegisteredTasksResult> {
  return getCache().get<RegisteredTasksResult>("registered-tasks");
}

export async function getCachedQueueDetails(): Promise<QueueDetailsResult> {
  return getCache().get<QueueDetailsResult>("queue-details");
}

File: services/celery-gateway/src/celery_gateway/services/cache.py
from __future__ import annotations

import asyncio
import logging
import time
from typing import Any, Callable, Coroutine, TypeVar

from .inspect_cache import InspectCache
from .redis_client import get_redis

logger = logging.getLogger(__name__)

T = TypeVar("T")


class _CacheEntry:
    __slots__ = ("data", "updated_at", "ttl_s", "task", "inflight", "refresh_fn")

    def __init__(
        self,
        ttl_s: float,
        refresh_fn: Callable[[], Coroutine[Any, Any, Any]],
    ) -> None:
        self.data: Any = None
        self.updated_at: float = 0.0
        self.ttl_s = ttl_s
        self.task: asyncio.Task[None] | None = None
        self.inflight: asyncio.Task[Any] | None = None
        self.refresh_fn = refresh_fn


class CeleryCache:
    def __init__(self, inspect_cache: InspectCache) -> None:
        self._inspect_cache = inspect_cache
        self._entries: dict[str, _CacheEntry] = {}

        self._register("active-tasks", 2.0, self._refresh_active_tasks)
        self._register("queue-depths", 30.0, self._refresh_queue_depths)
        self._register("task-history", 10.0, self._refresh_task_history)
        self._register("worker-inspect", 15.0, self._refresh_worker_inspect)
        self._register("registered-tasks", 60.0, self._refresh_registered_tasks)
        self._register("queue-details", 5.0, self._refresh_queue_details)

    def _register(
        self,
        key: str,
        ttl_s: float,
        refresh_fn: Callable[[], Coroutine[Any, Any, Any]],
    ) -> None:
        self._entries[key] = _CacheEntry(ttl_s, refresh_fn)

    async def get(self, key: str) -> Any:
        entry = self._entries.get(key)
        if entry is None:
            raise KeyError(f"Unknown cache key: {key}")

        if entry.updated_at == 0.0:
            await self._refresh(key)
            self._start_timer(key)

        if entry.data is None:
            return {}
        return entry.data

    async def _refresh(self, key: str) -> None:
        entry = self._entries[key]

        if entry.inflight is not None:
            try:
                await entry.inflight
            except Exception:
                pass
            return

        task = asyncio.create_task(entry.refresh_fn())
        entry.inflight = task

        try:
            data = await task
            entry.data = data
            entry.updated_at = time.monotonic()
        except Exception:
            logger.warning("Cache refresh failed for %r", key, exc_info=True)
        finally:
            entry.inflight = None

    def _start_timer(self, key: str) -> None:
        entry = self._entries[key]
        if entry.task is not None:
            return

        async def _loop() -> None:
            while True:
                await asyncio.sleep(entry.ttl_s)
                try:
                    await self._refresh(key)
                except Exception:
                    logger.warning(
                        "Background refresh timer failed for %r", key, exc_info=True
                    )

        entry.task = asyncio.create_task(_loop())

    def stop(self) -> None:
        for entry in self._entries.values():
            if entry.task is not None:
                entry.task.cancel()
                entry.task = None

    # ------------------------------------------------------------------
    # Refresh functions
    # ------------------------------------------------------------------

    async def _refresh_active_tasks(self) -> list[dict[str, Any]]:
        try:
            active_data, reserved_data = await asyncio.gather(
                self._inspect_cache.get("active"),
                self._inspect_cache.get("reserved"),
            )
            tasks: list[dict[str, Any]] = []
            seen_ids: set[str] = set()

            # Active tasks (currently executing)
            for worker_name, task_list in active_data.items():
                for t in task_list:
                    task_id = t.get("id", "")
                    if task_id:
                        seen_ids.add(task_id)
                    tasks.append({
                        "taskId": task_id,
                        "name": t.get("name", "unknown"),
                        "worker": worker_name,
                        "startedAt": t.get("time_start") or time.time(),
                        "status": "started" if t.get("acknowledged") else "received",
                        "args": str(t.get("args")) if t.get("args") is not None else None,
                        "kwargs": str(t.get("kwargs")) if t.get("kwargs") is not None else None,
                    })

            # Reserved tasks (received by worker, waiting to execute)
            for worker_name, task_list in reserved_data.items():
                for t in task_list:
                    task_id = t.get("id", "")
                    if task_id in seen_ids:
                        continue
                    seen_ids.add(task_id)
                    tasks.append({
                        "taskId": task_id,
                        "name": t.get("name", "unknown"),
                        "worker": worker_name,
                        "startedAt": t.get("time_start") or time.time(),
                        "status": "received",
                        "args": str(t.get("args")) if t.get("args") is not None else None,
                        "kwargs": str(t.get("kwargs")) if t.get("kwargs") is not None else None,
                    })

            return tasks
        except Exception:
            try:
                from .celery_redis import get_active_tasks
                return await get_active_tasks()
            except Exception:
                return []

    async def _refresh_queue_depths(self) -> dict[str, int]:
        try:
            data = await self._inspect_cache.get("active_queues")
            queue_names: set[str] = set()
            for _worker, queues_list in data.items():
                for q in queues_list:
                    name = q.get("name", q) if isinstance(q, dict) else str(q)
                    queue_names.add(name)
            if not queue_names:
                return {"celery": 0}
            from .celery_redis import get_queue_lengths
            return await get_queue_lengths(list(queue_names))
        except Exception:
            try:
                from .celery_redis import get_queue_lengths
                return await get_queue_lengths(["celery"])
            except Exception:
                return {"celery": 0}

    async def _refresh_task_history(self) -> list[dict[str, Any]]:
        try:
            from .celery_redis import get_historical_tasks
            tasks = await get_historical_tasks(50)

            names = [t["name"] for t in tasks if t.get("name") and t["name"] != "unknown"]
            if names:
                try:
                    redis = get_redis()
                    await redis.sadd("celeryhub:known-tasks", *names)
                except Exception:
                    pass

            return tasks
        except Exception:
            return []

    async def _refresh_worker_inspect(self) -> dict[str, Any] | None:
        try:
            methods = ["active", "registered", "stats", "active_queues"]
            results = await asyncio.gather(
                *[self._inspect_cache.get(m) for m in methods]
            )
            data: dict[str, Any] = {}
            for method, result in zip(methods, results):
                data[method] = result

            import datetime
            data["timestamp"] = datetime.datetime.now(datetime.timezone.utc).isoformat()
            data["cached"] = True
            return data
        except Exception:
            return None

    async def _refresh_registered_tasks(self) -> dict[str, Any]:
        try:
            from .celery_redis import get_known_task_names
            known_names = await get_known_task_names()

            by_worker: dict[str, list[str]] = {}
            try:
                registered = await self._inspect_cache.get("registered")
                if registered:
                    for worker_name, task_list in registered.items():
                        by_worker[worker_name] = sorted(task_list)
                    # Seed the known-tasks set with what workers actually
                    # register, so newly deployed tasks appear everywhere
                    # without waiting for their first event.
                    fresh = {
                        t for tasks in registered.values() for t in tasks
                        if t and not t.startswith("celery.")
                    }
                    new_names = fresh - set(known_names)
                    if new_names:
                        from .redis_client import get_redis

                        await get_redis().sadd(
                            "celeryhub:known-tasks", *new_names
                        )
                        known_names = sorted(set(known_names) | new_names)
            except Exception:
                pass

            all_tasks: set[str] = set(known_names)
            for tasks in by_worker.values():
                all_tasks.update(tasks)

            return {"byWorker": by_worker, "tasks": sorted(all_tasks)}
        except Exception:
            return {"byWorker": {}, "tasks": []}

    async def _refresh_queue_details(self) -> dict[str, Any]:
        try:
            depths = await self.get("queue-depths")
        except Exception:
            depths = {"celery": 0}

        queue_names = list(depths.keys())
        pending: dict[str, list[dict[str, str]]] = {}

        from .celery_redis import get_pending_tasks

        results = await asyncio.gather(
            *[get_pending_tasks(q, 200) for q in queue_names],
            return_exceptions=True,
        )

        for i, q in enumerate(queue_names):
            result = results[i]
            pending[q] = result if isinstance(result, list) else []

        return {
            "queueNames": queue_names,
            "depths": depths,
            "pending": pending,
        }
